fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rewrite/DictMappingRewriter.java

```diff
@@ -0,0 +1,207 @@
+    ScalarOperator rewriteAsDictMapping(ScalarOperator scalarOperator, Type type) {
+        final ColumnRefSet usedColumns = scalarOperator.getUsedColumns();
+        Preconditions.checkState(usedColumns.cardinality() == 1);
+        final Integer dictColumnId = decodeContext.stringColumnIdToDictColumnIds.get(usedColumns.getFirstId());
+        ColumnRefOperator dictColumn = decodeContext.columnRefFactory.getColumnRef(dictColumnId);
+        scalarOperator = new DictMappingOperator(dictColumn, scalarOperator.clone(), type);
+        return scalarOperator;
+    }
+
+    private class Rewriter extends ScalarOperatorVisitor<ScalarOperator, RewriterContext> {
+
+        ScalarOperator rewriteForScalarOperator(ScalarOperator operator, RewriterContext context) {
+            final ColumnRefSet usedColumns = operator.getUsedColumns();
+            // without intersect column, skip rewrite
+            if (!context.stringColumnSet.isIntersect(usedColumns)) {
+                context.hasAppliedOperator = false;
+                return operator;
+            }
+
+            if (usedColumns.cardinality() == 1) {
+                // check child support opt
+                List<ScalarOperator> children = Lists.newArrayList(operator.getChildren());
+                boolean hasApplied = false;
+                boolean disableApplied = false;
+                for (int i = 0; i < children.size(); i++) {
+                    context.reset();
+                    children.set(i, children.get(i).accept(this, context));
+                    hasApplied = hasApplied || context.hasAppliedOperator;
+                    disableApplied = disableApplied || context.hasUnsupportedOperator;
+                }
+                Preconditions.checkState(hasApplied);
+                if (!disableApplied) {
+                    context.hasAppliedOperator = true;
+                    return operator;
+                } else {
+                    context.hasAppliedOperator = false;
+                    return visit(operator, context);
+                }
+            } else {
+                // add dict operator for each child
+                return visit(operator, context);
+            }
+        }
+
+        @Override
+        public ScalarOperator visit(ScalarOperator scalarOperator, RewriterContext context) {
+            List<ScalarOperator> children = Lists.newArrayList(scalarOperator.getChildren());
+            boolean hasApplied = false;
+            boolean disableApplied = context.hasUnsupportedOperator;
+            // For any expression that does not support low-base optimization,
+            // if child already uses optimization, we need to add a DictExpr
+            for (int i = 0; i < children.size(); i++) {
+                context.reset();
+                ScalarOperator child = scalarOperator.getChild(i).accept(this, context);
+                // wrapper using DictExpr
+                if (context.hasAppliedOperator) {
```

Comment:
it is quite confusing, can we use another flags to indicate here we need to interpolate a DictMappingOperator？

Reply:
When we call the function "visit(ScalarOperator scalarOperator", this means that this current operator no longer supports using dictionary optimization, and at this time we need to block the dictionary optimization if child has already applied it. So at this time we need to rewrite the child using DictExpr in time

Reply:
I see, however, so can we use another name? for an example,  addDictExprToBlockDictOpt?